Merge namespaced key/values into non-namespaced key/values

jperras opened the pull request:
## Summary

Previously, if namespaced paths were used for Parameter Store hierarchical configuration, the entire non-namespaced configuration would need to be copied over.

This PR changes the behaviour so that namespaced configurations are merged into the equivalent non-namespaced configuration, which will allow for a more intelligent per-parameter override when necessary instead of requiring full parameter hierarchy duplication for every namespace.

For example, if you have the following keys:

``` text
/app/qa/database/host
/app/qa/database/port
/app/qa/database/password
/app/qa/pr-123/database/host
/app/qa/pr-123/database/password
```

The new behaviour will ensure that the end result of the namespaced configuration loading will include ``database.port``, which previously it would have not.

## Additional Details

- Refactor parameter loading logic into more composable units.
- Refactors test suite.
- Enforce prefix/namespace prefix sourcing from raw env.
- Adds an optional `NamespaceFilter` that can be used to ensure the merged result does not contain dangling namespace keys with their associated data (which was already merged down into the non-namespaced results).

jperras committed:
chore: Add pull_from_env_or_obj utility function

jperras committed:
feat: Merge non-namespaced params with namespaced params

- Refactor parameter loading into more composable units.
- Refactors test suite.
- Enforce prefix/namespace prefix sourcing from raw env.

jperras committed:
chore(docs): Update README.rst

jperras committed:
fix(tests): Add missing cleanup for parameter deletion

jperras committed:
feat: Add NamespaceFilter

When using namespaces, you can optionally provided a filter strategy to
ensure that the merged result does not contain the namespace key and
associated data.

jnhmcknight reviewed (commented):
Looks good. I would think that it just needs a little documentation about the filters, and consistency in the naming between `NAMESPACE_FILTER_STRATEGY` and `AWS_SSM_PARAMETER_NAMESPACE_FILTER`. (I don't see the latter used anywhere other than an error message, so not really sure what it should be doing.)

jperras commented:
@jnhmcknight Ah, thanks for flagging that. I'll update the documentation, and make the naming of those variables more consistent.

Do you think it's worth having a plugin prefix for the variables? e.g. instead of `AWS_SSM_<blah>`, we could do `DYNACONF_AWS_SSM_LOADER_<blah>` as a prefix to all variables that the plugin cares about.

jperras committed:
refactor: Unify environment parameter naming scheme

This used to be a mismatched naming convention (or lack thereof), which
was confusing at best.

jperras committed:
chore(docs): More details on configuration variables

jperras committed:
chore: Change logger name and use warn instead of exception level
